### DISCUSSION TITLE
[LGTM - C++] Compiling with C++20

### DISCUSSION THREAD

**Discussion by php1ic** [github/securitylab#535] [Q&A] [answered]
My project uses C++20 so requires a newer compiler than is available on the version of ubuntu (Eoan/19.10) used by LGTM. Is it recommended to use the built-in CodeQL with GitHub actions in this scenario?

I tried adding a PPA but they [aren't allowed](https://lgtm.com/help/lgtm/analysis-faqs#whats-the-default-build-environment-for-cpp-projects).

The latest logs from my attempts at coverage are here if needed.
https://lgtm.com/logs/fc06a617fed09f06ab0f25c5459b7199ec3dd11a/lang:cpp

**Comment by edoardopirovano** [comment, marked as answer]
Greetings, thanks for getting in touch with us about this! We're actually in the process of updating the default version of Ubuntu used by LGTM to 20.04 and, unless any unforeseen problems arise in our testing, expect to have that rolled out later this week. It may well be that your project can be successfully built after we've completed that upgrade, so I would recommend trying again next week.

Having said that, I will note that support for C++20 in CodeQL is currently in beta (see https://help.semmle.com/QL/ql-support/language-support/), so while things should more-or-less work you may notice some issues, particularly when using new features of C++20. Do get in touch if you encounter any such issues.

**Reply by php1ic** [reply]
Thanks @edoardopirovano ! I'll try again next week.

Is there any plans to allow PPAs, or even just enable the [ubuntu-toolchain](https://launchpad.net/~ubuntu-toolchain-r/+archive/ubuntu/test) by default to give access to newer versions of gcc?

**Reply by edoardopirovano** [reply]
There's no plans for that as far as I'm aware, no. We'll certainly take it into consideration for future work, though.

However, we've mostly shifted our focus off LGTM and towards code scanning on GitHub Actions, so it is unlikely we will be making major changes to LGTM at the moment.

**Reply by php1ic** [reply]
OK, good to know and thanks for the response.

**Comment by jbj** [comment]
The upgrade of the build environment to Ubuntu 20.04 was deployed on February 9. That upgrade allows you to have GCC 10 since Ubuntu has [an official package for that](https://packages.ubuntu.com/focal/g++-10). There is no GCC 11 package, so your use of C++20 on LGTM is limited to what's supported by GCC 10 (and by CodeQL).

I triggered [a test run of your project with GCC 10](https://lgtm.com/logs/5425b89d7dac1f62e33f8429da44e3a7b8f4a31e/lang:cpp), and the logs show compilation errors related to `string_view`. I don't know if there's any possible workaround you can apply.

**Reply by php1ic** [reply]
Thanks @jbj. I make heavy use of [std::from_chars](https://en.cppreference.com/w/cpp/utility/from_chars) so need at least GCC 11. The project [does compile](https://github.com/php1ic/nuclear-data-reader/runs/5266768366?check_suite_focus=true) with clang 12 on GitHub actions, however, it [doesn't compile](https://lgtm.com/logs/c77f315da86d0832ffb314b7ada7b496661c55fd/lang:cpp) on the LGTM build environment.

The main difference seems to be the additional PPA which provides GCC 11 and what looks to be the necessary header(s).
